### ISSUE TITLE
PoolingHttpClientConnectionManager not releasing connections or resources?

### ISSUE THREAD

**Issue by stephan-bosch** [SAP/cloud-sdk-java#261]
Greetings,

This is more of a question that might be a possible issue, ill provide more details if required..

Recently I did a memory analysis on our application and found that getting the destination is holding onto concurrent hash map data.
The reason i know this is from retrieving the destination is because it is the only location where io.vavr.control.Try$Success is used.

![image](https://github.com/SAP/cloud-sdk-java/assets/85499114/e947023e-e0a0-41b3-998f-bc3b590fe732)

it looks like the retaining has something to do with jackson 

![image](https://github.com/SAP/cloud-sdk-java/assets/85499114/b0c4ce80-2e92-4499-a596-4f702e5351b1)

![image](https://github.com/SAP/cloud-sdk-java/assets/85499114/d5520388-755e-4a71-b877-9855aa002ad3)

The way the destination is called is as follows:

![image](https://github.com/SAP/cloud-sdk-java/assets/85499114/b1b34160-60d0-4747-889c-5607e89276b1)

An sheduled job runs that invokes getDestination with the required tenant context and a callable object.

this is based on: 
 <spring-boot.version>2.6.4</spring-boot.version>
<sap.sdk.version>3.64.0</sap.sdk.version>
<java.version>1.8</java.version>

i am not that familiair with what is expected on retrieving the destination, but it seems excessive that so many concurrent hashmap nodes and connections are retained after the destination is retrieved.

i would like to know if this is expected behaviour, but it does take like 30% of the application heap memory.

kind regards,

Stephan

**Comment by newtork** [comment]
Hi Stephan,

Good finding; we take (reasonable) memory and performance optimization very serious!
However the version `3.64.0` of _SAP Cloud SDK_ that you mention  [is long deprecated.](https://sap.github.io/cloud-sdk/docs/java/release-policy)

We suggest to upgrade. Especially since I also noticed you're using Spring Boot 2 :)

Kind regards
Alexander

**Comment by stephan-bosch** [comment]
Hi newtork,

We have the to-be-planned production release that is closer up to date with <sap.sdk.version>4.27.0</sap.sdk.version>.
i suspect this will be at least a month away.

is it possible to reopen the issue once i have the production memory dump with the new version if it seems still persisting? 

But for me to benchmark, what is considered reasonable memory use?
And based on screenshots even though it's an older version, is that considered reasonable?

kind regards,

stephan

**Comment by newtork** [comment]
I mentioned "(reasonable)" because memory footprint can almost always be reduced. Some times at cost of API design / convenience. That would be unreasonable.

> is it possible to reopen the issue once i have the production memory dump with the new version if it seems still persisting?
> And based on screenshots even though it's an older version, is that considered reasonable?

Memory leaks like the one you found are not reasonable!
Fixing memory leaks  is high priority, We're happy about any complains to improve the product.

> But for me to benchmark, what is considered reasonable memory use?

 When objects are not garbage collected after usage.
**Note:** We maintain application-scoped caches. They may grow with different tenants / users / destination.

**Comment by stephan-bosch** [comment]
Hi newtork

Thank you for your reply,
As mentioned new release won't be in for a while (1-2 months), so this post will become stale till I get back to this. 

Feel free to close it or update this thread as you see fit.

kind regards,

Stephan

**Comment by newtork** [comment]
Hi Stephan,

You will be happy to hear we greatly improved performance and token/destination caching with recent versions.

**In SDK v4**, we added an optional destination change-detection mode with `ScpCfDestinationLoader.Cache.enableChangeDetection()`. It takes advantage of "cheap" calls to destination service to check whether a destination has changed, before making a "costly" call. Depending on number of destinations and active authentication type, it can save significant traffic / computation.

**In SDK v5**, we enabled this mode by default.

If you encounter any bottleneck with destination service interaction using (recent versions of) _SAP Cloud SDK_, we will be happy to learn about and fix it. For now I will close the ticket. Happy hacking!

Kind regards
Alexander
